File: Daily/287.find-the-duplicate-number.js
/*
 * @lc app=leetcode id=287 lang=javascript
 *
 * [287] Find the Duplicate Number
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function (nums) {
    let duplicate = -1;

    for (let num of nums) {
        num = Math.abs(num);
        if (nums[num - 1] < 0) {
            duplicate = num;
            break;
        }
        nums[num - 1] = -nums[num - 1];
    }

    for (let i = 0; i < nums.length; i++) {
        nums[i] = Math.abs(nums[i]);
    }

    return duplicate;
};
// @lc code=end

File: Daily/287.find-the-duplicate-number.cpp
/*
 * @lc app=leetcode id=287 lang=cpp
 *
 * [287] Find the Duplicate Number
 */
#include <cmath>
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int findDuplicate(vector<int>& nums) {
    int slow = 0;
    int fast = 0;
    while (true) {
      slow = nums[slow];
      fast = nums[nums[fast]];
      if (slow == fast)
        break;
    }

    int temp = 0;
    while (true) {
      slow = nums[slow];
      temp = nums[temp];
      if (slow == temp)
        break;
    }
    return slow;
  }
};
// @lc code=end